### PR TITLE
raft: implement fast log rejection

### PR DESCRIPTION
**Pull request by hicqu**
The fast log rejection algorithm is descripted in the Raft paper, which can reduce network round trips when peer with less logs becomes leader.

There are 2 cases that the PR can improve:

#### A. The conflict point can be found by leader quickly:
```
leader log   : (idx=100,term=2)(idx=101,term=5)[...](idx=120,term=5)
follower logs: (idx=100,term=2)(idx=101,term=4)[...](idx=110,term=4)
```
When leader appends entries from `index=120`, follower will reject it with `reject=110,term=4`. So leader can found the conflict position and then appends entries from `index=101`.

#### B. The conflict point can be found by follower quickly:
```
leader logs  : (idx=100,term=2)[...](idx=105,term=2)
follower logs: (idx=100,term=2)(idx=101,term=2)(idx=102,term=4)
```
When leader appends entries from `index=105`, follower will reject it with `reject=100,term=2` if it can found the conflict position. After that leader can continue to append entries from `index=101`.

Signed-off-by: qupeng <qupeng@pingcap.com>

Please read https://github.com/etcd-io/etcd/blob/master/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by hicqu** [comment]
Friendly ping @bdarnell @tbg .

**Comment by tbg** [comment]
For context, here is the passage from the thesis:

> If desired, the protocol can be optimized to reduce the number of rejected AppendEntries RPCs.
For example, when rejecting an AppendEntries request, the follower can include the term of the
conflicting entry and the first index it stores for that term. With this information, the leader can
decrement nextIndex to bypass all of the conflicting entries in that term; one AppendEntries RPC
will be required for each term with conflicting entries, rather than one RPC per entry

File: raft/raft.go [hunk 1]
```diff
@@ -1381,7 +1387,15 @@ func (r *raft) handleAppendEntries(m pb.Message) {
...
 			r.id, r.raftLog.zeroTermOnErrCompacted(r.raftLog.term(m.Index)), m.Index, m.LogTerm, m.Index, m.From)
-		r.send(pb.Message{To: m.From, Type: pb.MsgAppResp, Index: m.Index, Reject: true, RejectHint: r.raftLog.lastIndex()})
+		reject := r.raftLog.lastIndex()
+		if m.Index < reject {
+			reject = m.Index
+		}
+		logTerm, _ := r.raftLog.term(reject)
+		r.send(pb.Message{To: m.From, Type: pb.MsgAppResp, Index: m.Index, Reject: true,
+			RejectHint: reject,
+			LogTerm:    logTerm,
```
Comment (tbg) [inline raft/raft.go:1397]: Please add a comment on `LogTerm` (in the proto) that in a MsgAppResp, it is only set on rejections, and to the term corresponding to RejectHint. This is necessary because the naive assumption would be that it corresponds to `Index`.

File: raft/log.go [hunk 2]
```diff
@@ -138,6 +138,21 @@ func (l *raftLog) findConflict(ents []pb.Entry) uint64 {
 	return 0
 }
 
+func (l *raftLog) findConflictByTerm(index uint64, term uint64) uint64 {
```
[the remark this answers is not in the document]
Comment (tbg) [reply raft/log.go:141]: I'm not even sure that comment is correct. But you get the idea, explain the inputs and outputs and what it's for.

File: raft/log.go [hunk 3]
```diff
@@ -138,6 +138,21 @@ func (l *raftLog) findConflict(ents []pb.Entry) uint64 {
...
+	var conflictIndex uint64
```
Comment (tbg) [inline raft/log.go:142]: Why not:

```go
	for {
		if logTerm, err := l.term(index); err == nil && logTerm > term {
			index--
			continue
		}
		return index
	}
```

File: raft/raft_test.go [hunk 4]
```diff
@@ -4141,6 +4141,218 @@ func TestPreVoteMigrationWithFreeStuckPreCandidate(t *testing.T) {
...
+func TestFastLogRejection(t *testing.T) {
+	tests := []struct {
+		logs  []pb.Entry
+		ent1  []pb.Entry
+		ent2  []pb.Entry
+		term  uint64
+		index uint64
+	}{
+		// append reject, return index that is not conflict
+		{
```
Comment (tbg) [inline raft/raft_test.go:4153]: Please add narration to at least a few of the test cases, similar to how I did it in the comment suggestion. As is, these tests are inscrutable for anyone who isn't willing to stare at them for 20 minutes. That is not ideal.
I'm also curious whether you've considered using env-driven testing for this (raft/testdata). That tends to be much more user-friendly IMO.

File: raft/raft_test.go [hunk 5]
```diff
@@ -4141,6 +4141,218 @@ func TestPreVoteMigrationWithFreeStuckPreCandidate(t *testing.T) {
...
+			5,
+		},
+	}
+
+	for i, test := range tests {
+		s1 := NewMemoryStorage()
+		s1.Append(test.logs)
+
+		n1 := newTestRaft(1, []uint64{1, 2, 3}, 10, 1, s1)
+		n2 := newTestRaft(2, []uint64{1, 2, 3}, 10, 1, NewMemoryStorage())
```
Comment (tbg) [inline raft/raft_test.go:4295]: are you intentionally not appending `test.logs` on n2? I thought test.logs was a shared prefix and ents1 and ents2 would be possibly different suffixes. It doesn't seem to matter for this test but seems cleaner (less confusing) anyway.

File: raft/raft.go [hunk 6]
```diff
@@ -1122,9 +1122,15 @@ func stepLeader(r *raft, m pb.Message) error {
 		pr.RecentActive = true
 
 		if m.Reject {
-			r.logger.Debugf("%x received MsgAppResp(MsgApp was rejected, lastindex: %d) from %x for index %d",
-				r.id, m.RejectHint, m.From, m.Index)
+			r.logger.Debugf("%x received MsgAppResp(MsgApp was rejected, lastindex: %d) from %x for index %d, log term %d",
+				r.id, m.RejectHint, m.From, m.Index, m.LogTerm)
 			if pr.MaybeDecrTo(m.Index, m.RejectHint) {
```
Comment (tbg) [inline raft/raft.go:1127]: PS while we're here - the comment on `MaybeDecrTo` is wrong. It says that the first argument (`rejected`) is the index that we tried to append. But that's not true, we tried to append `rejected+1` (i.e. the `m.Index` of the MsgApp was `rejected`) and it failed due to a mismatch at `rejected` itself.

**Comment by hicqu** [comment]
@tbg thanks for your review! Here is a little change list about the latest commit:
1. Document `findConflictByTerm` correctly;
2. Comment `logTerm` when it's filled into `MsgAppResp` and read out from `MsgAppResp`;
3. Simplify test cases;
4. Document `logTerm` in proto file.

And, I have updated the PR's decription to show which cases can be improved by this PR.

**Comment by hicqu** [comment]
Friendly ping @tbg .

**Comment by hicqu** [comment]
Could you take a look again? @tbg

**Comment by xiang90** [comment]
@hicqu can you resolve the merging conflicts? i will go over it again next week.

File: raft/raft_test.go [hunk 7]
```diff
@@ -4141,6 +4141,170 @@ func TestPreVoteMigrationWithFreeStuckPreCandidate(t *testing.T) {
...
+		}
+
+		if n2.Step(msgs[0]) != nil {
+			t.Errorf("peer 2 step append fail")
+		}
+		msgs = n2.readMessages()
+		if len(msgs) != 1 {
+			t.Errorf("can't read 1 message from peer 2")
+		}
+		if msgs[0].Type != pb.MsgAppResp {
```
Comment (nvanbenschoten) [inline raft/raft_test.go:4291]: Do we want to assert on the contents of this `MsgAppResp`? At a minimum, we should check that its Reject field is true. But given the complex follower-side searching we do, it would also be interesting to test the `RejectHint` and `LogTerm` for each test case.

File: raft/log.go [hunk 8]
```diff
@@ -138,6 +138,19 @@ func (l *raftLog) findConflict(ents []pb.Entry) uint64 {
 	return 0
 }
 
+// findConflictByTerm takes an (index, term) pair (indicating a conflicting log
+// entry on a follower during an append) and finds the last index for that term
+// (or the first index of the log, whichever is larger).
+func (l *raftLog) findConflictByTerm(index uint64, term uint64) uint64 {
+	for {
+		if logTerm, err := l.term(index); err == nil && logTerm > term {
```
Comment (nvanbenschoten) [inline raft/log.go:146]: The `err == nil` case is concerning me. Do we have test cases that exercise it?

At a minimum, I think it would be valuable to create a `firstIndex := err != nil` variable to hint to readers what that error means in this context. And then structure this all as:
```go
func (l *raftLog) findLastConflictingIndexAtOrBelowTerm(index uint64, searchTerm uint64) uint64 {
    for {
            logTerm, err := l.term(index)
            firstIndex := err != nil
            if logTerm <= searchTerm || firstIndex {
                break
            }
            index--
    }
    return index
}
```

**Review by nvanbenschoten — COMMENTED** [review]
Hi @hicqu,

Sorry to let this PR sit. It still seems like a good idea, as it should minimize probing time after leader elections and prevent long uncommitted log tails from causing problems. Are you still interested in pushing this change over the finish line?

**Comment by tbg** [comment]
I'll rebase this - given that Nathan and I are both engaging now and are coworkers, I think it's reasonable that the two of us just take over this PR and get it over the finish line.

**Comment by tbg** [comment]
Rebased. I didn't address any review comments yet but it should pass tests.

File: raft/log.go (hunk 8 above)
Comment (nvanbenschoten) [reply raft/log.go:146]: Done.

File: raft/raft_test.go (hunk 7 above)
Comment (nvanbenschoten) [reply raft/raft_test.go:4417]: Done.

File: raft/raft.go [hunk 9]
```diff
@@ -1381,7 +1400,19 @@ func (r *raft) handleAppendEntries(m pb.Message) {
 	} else {
 		r.logger.Debugf("%x [logterm: %d, index: %d] rejected MsgApp [logterm: %d, index: %d] from %x",
 			r.id, r.raftLog.zeroTermOnErrCompacted(r.raftLog.term(m.Index)), m.Index, m.LogTerm, m.Index, m.From)
-		r.send(pb.Message{To: m.From, Type: pb.MsgAppResp, Index: m.Index, Reject: true, RejectHint: r.raftLog.lastIndex()})
+		reject := r.raftLog.lastIndex()
+		if m.Index <= reject {
+			reject = r.raftLog.findConflictByTerm(m.Index, m.LogTerm)
```
[the remark this answers is not in the document]
Comment (nvanbenschoten) [reply raft/raft.go:1385]: Done.

File: raft/raft.go [hunk 10]
```diff
@@ -1122,9 +1122,28 @@ func stepLeader(r *raft, m pb.Message) error {
...
+			r.logger.Debugf("%x received MsgAppResp(MsgApp was rejected, hint: %d) from %x for index %d, log term %d",
+				r.id, m.RejectHint, m.From, m.Index, m.LogTerm)
+			// The reject hint is the suggested next base entry for appending,
+			// i.e. hint+1 would be sent as the next append. 'hint' is usually the
+			// last index of the follower, or the base index (m.Index) from the
+			// append being responsed to.
+			hint := m.RejectHint
+			if m.LogTerm > 0 {
+				// LogTerm is the term that the follower has at index RejectHint. If the
+				// follower has an uncommitted log tail, we would end up probling one by
```
[the remark this answers is not in the document]
Comment (nvanbenschoten) [reply raft/raft.go:1118]: Done.

**Comment by nvanbenschoten** [comment]
@tbg I addressed my suggestions from earlier and added some more testing, including a nice interactive test that demonstrates the improvements of this change on the log configuration in Figure 7 of the raft paper:

Screen Shot 2021-02-10 at 1 11 04 AM


I'm not a maintainer of this repo, so I wasn't able to push changes directly to this PR. So they're over in https://github.com/etcd-io/etcd/compare/master...nvanbenschoten:fast-log-rejection?expand=1.

**Comment by tbg** [comment]
@nvanbenschoten I added an extra commit to explain things better (fallout from explaining it to myself), PTAL

File: raft/raft.go [hunk 11]
```diff
@@ -1106,32 +1106,93 @@ func stepLeader(r *raft, m pb.Message) error {
...
+				// There is a similar mechanism on the follower (implemented in
+				// handleAppendEntries via a call to findConflictByTerm) that is
+				// useful if the follower has a large divergent uncommitted log
+				// tail, as in this example:
+				//
+				//   idx        1 2 3 4 5 6 7 8 9
+				//              -----------------
+				//   term (L)   1 3 3 3 3 3 3 3 7
+				//   term (F)   1 3 3 4 4 5 5 5 6
+				//   idx
```
Comment (nvanbenschoten) [inline raft/raft.go:1174]: Was this line intentional?

File: raft/raft.go [hunk 12]
```diff
@@ -1106,32 +1106,93 @@ func stepLeader(r *raft, m pb.Message) error {
...
+				// At index 1, however, the leader can draw no such conclusion,
+				// as its term 1 is not larger than the term 2 from the
+				// follower's rejection. We thus probe at 1, which will succeed
+				// in this example. In general, with this approach we probe at
+				// most once per term found in the leader's log.
+				//
+				// There is a similar mechanism on the follower (implemented in
+				// handleAppendEntries via a call to findConflictByTerm) that is
+				// useful if the follower has a large divergent uncommitted log
+				// tail, as in this example:
```
Comment (nvanbenschoten) [inline raft/raft.go:1168]: Consider mentioning that this is only useful if the follower's large divergent tail has terms that are larger than those at the corresponding indexes in the leader's log. In the process of understanding this issue, I've had to convince myself of that at least three times. I've also had to convince myself that this situation is even possible, which it is.

File: raft/raft.go [hunk 13]
```diff
@@ -1106,32 +1106,93 @@ func stepLeader(r *raft, m pb.Message) error {
...
+				// and many preceding indexes; the leader would have to probe linearly
+				// until it would finally hit index 3, where the probe would succeed.
+				//
+				// Instead, we apply a similar optimization on the follower. When the
+				// follower receives the probe at index 8 (log term 3), it concludes
+				// that all of the leader's log preceding that index has log terms of
+				// 3 or below. The largest index in the follower's log with a log term
+				// of 3 or below is index 3. The follower will thus return a rejection
+				// for index=3, log term=3 instead. The leader's next probe will then
+				// succeed.
```
Comment (nvanbenschoten) [inline raft/raft.go:1192]: Consider spelling out what this next probe will be.

**Review by nvanbenschoten — APPROVED** [review]
Nice improvements to the comments. They make this much easier to understand.

**Comment by tbg** [comment]
I'm seeing the functional tests fail here for the x-th time, last failure [here](https://travis-ci.com/github/etcd-io/etcd/builds/216953736). These logs make it really hard to parse what's going on (what test is this even running?) but it passes locally:

```
docker run --rm       --volume=`pwd`:/go/src/go.etcd.io/etcd gcr.io/etcd-development/etcd-test:go${TRAVIS_GO_VERSION}       /bin/bash -c "GO_BUILD_FLAGS='-v -mod=readonly' ./build && GOARCH=amd64 PASSES='functional' ./test"
[...]
'etcd-tester' succeeded
./test.sh: line 174: kill: (5011) - No such process
./test.sh: line 174: kill: (5012) - No such process
./test.sh: line 174: kill: (5013) - No such process
functional test PASS!
'functional' completed at Mon Feb 15 08:52:32 UTC 2021
SUCCESS
```

I would blame it on this target being flaky, but I see that while master is pretty red, the functional suite hasn't failed in any of the last five iterations at the time of writing: https://travis-ci.com/github/etcd-io/etcd/branches

@ptabor do you have any insights here?

**Comment by tbg** [comment]
```
stderr: {"level":"error","ts":1613151559.1980429,"caller":"tester/cluster_run.go:50","msg":"round FAIL","round":0,"case":10,"case-total":22,"error":"recovery error: rpc error: code = Unknown desc = write tcp 127.0.0.1:2381->127.0.0.1:45082: write: connection reset by peer","stacktrace":"go.etcd.io/etcd/tests/v3/functional/tester.(*Cluster).Run

	/go/src/go.etcd.io/etcd/tests/functional/tester/cluster_run.go:50

main.main

	/go/src/go.etcd.io/etcd/tests/functional/cmd/etcd-tester/main.go:59

runtime.main

	/usr/local/go/src/runtime/proc.go:204"}
```

```
stderr: {"level":"error","ts":1613047421.8999214,"caller":"tester/cluster_run.go:50","msg":"round FAIL","round":0,"case":9,"case-total":22,"error":"injection error: cluster too slow: only 133 commits in 6 retries","stacktrace":"go.etcd.io/etcd/tests/v3/functional/tester.(*Cluster).Run

	/go/src/go.etcd.io/etcd/tests/functional/tester/cluster_run.go:50

main.main

	/go/src/go.etcd.io/etcd/tests/functional/cmd/etcd-tester/main.go:59

runtime.main

	/usr/local/go/src/runtime/proc.go:204"}
```

**Comment by tbg** [comment]
Managed to reproduce locally (4/5 passed, then this):

```
{"level":"info","ts":1613382879.2866938,"caller":"agent/handler.go:41","msg":"handler success","operation":"BLACKHOLE_PEER_PORT_TX_RX"}
stderr: {"level":"info","ts":1613382879.286739,"caller":"tester/cluster.go:470","msg":"received response","operation":"BLACKHOLE_PEER_PORT_TX_RX
","from":"127.0.0.1:1379","success":true,"status":"blackholed peer port tx/rx"}
stderr: {"level":"info","ts":1613382879.286763,"caller":"tester/case.go:252","msg":"trigger snapshot START","desc":"BLACKHOLE_PEER_PORT_TX_RX_LE
ADER_UNTIL_TRIGGER_SNAPSHOT","etcd-snapshot-count":2000}
stderr: {"level":"info","ts":1613382879.4084778,"caller":"tester/case.go:307","msg":"trigger snapshot RETRY","retries":0,"committed-entries":0,"
etcd-snapshot-count":2000,"start-revision":101708,"last-revision":101708,"took":0.121712851}
stderr: {"level":"info","ts":1613382880.4792778,"caller":"tester/case.go:307","msg":"trigger snapshot RETRY","retries":1,"committed-entries":0,"
etcd-snapshot-count":2000,"start-revision":101708,"last-revision":101708,"took":1.192513894}
stderr: {"level":"info","ts":1613382881.5366092,"caller":"tester/case.go:307","msg":"trigger snapshot RETRY","retries":2,"committed-entries":337
,"etcd-snapshot-count":2000,"start-revision":101708,"last-revision":102045,"took":2.249845412}
stderr: {"level":"info","ts":1613382882.599872,"caller":"tester/case.go:307","msg":"trigger snapshot RETRY","retries":3,"committed-entries":762,
"etcd-snapshot-count":2000,"start-revision":101708,"last-revision":102470,"took":3.313107066}
stderr: {"level":"info","ts":1613382883.6700158,"caller":"tester/case.go:307","msg":"trigger snapshot RETRY","retries":4,"committed-entries":121
4,"etcd-snapshot-count":2000,"start-revision":101708,"last-revision":102922,"took":4.383252135}
stderr: {"level":"info","ts":1613382884.7321687,"caller":"tester/case.go:307","msg":"trigger snapshot RETRY","retries":5,"committed-entries":163
3,"etcd-snapshot-count":2000,"start-revision":101708,"last-revision":103341,"took":5.445403496}
stderr: {"level":"error","ts":1613382885.73243,"caller":"tester/cluster_run.go:50","msg":"round FAIL","round":0,"case":11,"case-total":22,"error
":"injection error: cluster too slow: only 1633 commits in 6 retries","stacktrace":"go.etcd.io/etcd/tests/v3/functional/tester.(*Cluster).Run
        /go/src/go.etcd.io/etcd/tests/functional/tester/cluster_run.go:50
main.main
        /go/src/go.etcd.io/etcd/tests/functional/cmd/etcd-tester/main.go:59
runtime.main
        /usr/local/go/src/runtime/proc.go:204"}

**Comment by tbg** [comment]
I managed to reproduce a similar flake on `master` within three runs:

```
stderr: {"level":"error","ts":1613391270.6197243,"caller":"tester/cluster_run.go:50","msg":"round FAIL","round":0,"case":10,"case-total":22,"err
or":"injection error: cluster too slow: only 1100 commits in 6 retries","stacktrace":"go.etcd.io/etcd/tests/v3/functional/tester.(*Cluster).Run
        /go/src/go.etcd.io/etcd/tests/functional/tester/cluster_run.go:50
main.main
        /go/src/go.etcd.io/etcd/tests/functional/cmd/etcd-tester/main.go:59
runtime.main
        /usr/local/go/src/runtime/proc.go:204"}
stderr: {"level":"info","ts":1613391270.6198068,"caller":"tester/cluster_run.go:342","msg":"closing stressers before archiving failure data","ro
und":0,"case":10,"case-total":22}
```

Going to merge this PR then, as it is unlikely that the failure is causally related to this PR.
